nixos/matrix-synapse: Add UNIX domain socket listener support

mweinelt opened the pull request:
Exposes two options, `path` and `mode`, to configure the location and permissions on the socket file.

Adds an assertion, that either `path` or `bind_addresses` and `port` are configured on every listener.

Migrates the default replication listener of the main instance to a UNIX domain socket, because it is more efficient.

cc https://github.com/NixOS/nixos-org-configurations/pull/336

## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

mweinelt committed:
nixos/matrix-synapse: add UNIX domain socket listener support

Exposes two options, `path` and `mode`, to configure the location and
permissions on the socket file.

The `mode` needs to be specified as string in octal and will be converted
into a decimal integer, so it correctly passes through the YAML parser
and arrives at the `os.chmod` call in the Twisted codebase. What a fun
detour.

Adds an assertion, that either `path` or `bind_addresses` and `port` are
configured on every listener.

Migrates the default replication listener of the main instance to a UNIX
domain socket, because it is more efficient.

Introduces the `enableRegistrationScript` option, to gracefully disable
the user registration script, when the client listener listens on a UNIX
domain socket, which is something the script does not support.

999eagle reviewed (approved):
Otherwise LGTM. I've deployed this on my server now and it works fine. Had to manually change the replication listener and instance map to use the new path though because I'm setting those options myself.

999eagle reviewed (changes requested):
Okay no, I found a big issue with the `mode` setting. When not changing the mode from its default of 660 (which is 1224 in octal), the socket is created with `s-w--w-r-T` permissions which is completely wrong (but apparently enough for replication to still work?). I had to set the mode to 438 (which is 666 in octal) to get `srw-rw-rw-` permissions on the socket.
Apparently, Synapse interprets the mode as octal even if it's a (decimal) integer, so we would need to serialize this as `mode: 0o660` or something similar in the generated yaml file.

EDIT: also the final `homeserver.yaml` still contains `port: null` for all UDS listeners and `mode: 660`,`path: null` for all TCP listeners.

mweinelt commented:
Thanks for testing.

I initially had the mode as `nullOr str`, but `os.chmod` complained about that. So that's when I made it `nullOr int`. Not sure how to represent file permissions properly in NixOS modules.

I'm a bit surprised, that the final config still has `null` values, since the renderer filters out `null` values. Have to recheck the code path.

Ma27 commented:
> I initially had the mode as nullOr str, but os.chmod complained about that. So that's when I made it nullOr int. Not sure how to represent file permissions properly in NixOS modules.

Because it had quotes in the final config and was thus interpreted as a string? Would kinda suck though if that got propagated all the way down to the chmod call.

How about `types.coercedTo` or something like this instead? Apparently YAML also accepts `0X` as base8, perhaps we can make use of that somehow.

>  Apparently, Synapse interprets the mode as octal even if it's a (decimal) integer, so we would need to serialize this as mode: 0o660 or something similar in the generated yaml file.

I don't really understand where this happens though, I skimmed a bit through the code of synapse and twisted and to me it seems that no more conversion hacks are done, so apparently a correct representation is assumed.

Anyways, I think it's wrong that we even have to care and I think it's a bad idea to use a base10 integer when we need a base8 one. I think that a string with coercion (and a very strict regex check) is a better option here.

Ma27 commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -992,6 +1025,17 @@ in {
+        message = ''
+          Listener configurations require either `bind_addresses` and `port` for TCP sockets,
+          or `path` for UNIX domain sockets.
+        '';
```
It's perhaps a little easier to generate one assertion for each listener and then dump the `path` & `bind_addresses` into the message. That probably makes it easier to spot the bogus listener.

mweinelt replied:
Added an assertion, and mentioned the path. Since `bind_addresses`, `port` and `tls` are unsupported when `path` is set, and any one of them could be null, I decided not to print them.

mweinelt commented:
This is the code path up to the `os.chmod` call in twisted.

```pytb
Error during startup:
Traceback (most recent call last):
  File "/nix/store/gxma1wp6xc25vp8lza0bkhm30abrl1l1-matrix-synapse-1.100.0/lib/python3.11/site-packages/synapse/app/_base.py", line 258, in wrapper
    await cb(*args, **kwargs)
  File "/nix/store/gxma1wp6xc25vp8lza0bkhm30abrl1l1-matrix-synapse-1.100.0/lib/python3.11/site-packages/synapse/app/homeserver.py", line 370, in start
    await _base.start(hs)
  File "/nix/store/gxma1wp6xc25vp8lza0bkhm30abrl1l1-matrix-synapse-1.100.0/lib/python3.11/site-packages/synapse/app/_base.py", line 594, in start
    hs.start_listening()
  File "/nix/store/gxma1wp6xc25vp8lza0bkhm30abrl1l1-matrix-synapse-1.100.0/lib/python3.11/site-packages/synapse/app/homeserver.py", line 258, in start_listening
    self._listener_http(self.config, listener)
  File "/nix/store/gxma1wp6xc25vp8lza0bkhm30abrl1l1-matrix-synapse-1.100.0/lib/python3.11/site-packages/synapse/app/homeserver.py", line 147, in _listener_http
    ports = listen_http(
            ^^^^^^^^^^^^
  File "/nix/store/gxma1wp6xc25vp8lza0bkhm30abrl1l1-matrix-synapse-1.100.0/lib/python3.11/site-packages/synapse/app/_base.py", line 441, in listen_http
    ports = listen_unix(
            ^^^^^^^^^^^^
  File "/nix/store/gxma1wp6xc25vp8lza0bkhm30abrl1l1-matrix-synapse-1.100.0/lib/python3.11/site-packages/synapse/app/_base.py", line 388, in listen_unix
    cast(Port, reactor.listenUNIX(path, factory, backlog, mode, wantPID))
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/gmygp534ydqbs1iw63090m9vixkcydrq-python3-3.11.7-env/lib/python3.11/site-packages/twisted/internet/posixbase.py", line 265, in listenUNIX
    p.startListening()
  File "/nix/store/gmygp534ydqbs1iw63090m9vixkcydrq-python3-3.11.7-env/lib/python3.11/site-packages/twisted/internet/unix.py", line 416, in startListening
    os.chmod(self.port, self.mode)
TypeError: 'str' object cannot be interpreted as an integer
```

999eagle commented:
From what I've gathered the mode actually has to be an integer but its bits are directly set as the file mode. There is no special handling anywhere because python uses [an octal literal](https://github.com/element-hq/synapse/blob/50a332cf30a6b00e57964746a3c0786822985ab0/synapse/config/server.py#L1009) as default for the mode, so the default value is actually `0o666`, which is 438 in decimal.
The config parser does not seem to do any kind of type conversions so the expectation is that the yaml already contains an integer. I *think* yaml supports octal literals using the same `0o`-syntax, so that's how we should specify the mode.

Ma27 commented:
> There is no special handling anywhere because python uses an octal literal as default for the mode, so the default value is actually 0o666, which is 438 in decimal.

Yeah, can confirm that code-wise.

Rereading your previous comment, I think I misunderstood the "interprets it as octal" part and thought that you specified the decimal number which lead to the error not the other way round, sorry for that!

>  I think yaml supports octal literals using the same 0o-syntax, so that's how we should specify the mode.

Quickly checked that you can indeed write `0666` in YAML to get a base8 int, so that's probably what we'll want to output.

But without being able to express octal numbers in Nix (and I'm not _entirely_ sure if I'll like it then) I strongly dislike the `mode` to be a number, so I stand by my suggestion to use a string with coercedTo and strict validation.

mweinelt commented:
My understanding is as follows:
  - Nix basically supports only base 10 integers and strings.
  - Synapse uses pyyaml to read the config.
  - pyyaml only supports YAML 1.1, where octal numbers need to be specified with leading zeros.
  - Python's `os.chmod` only supports integers, but respects different representations, like `0o660`
  - I don't think there is a way to represent an int in Nix, that has leading zeros.
  - I don't understand how coercing the type can save us here

I see two possible solutions:
- Hardcode the option in its decimal form
- Write a function that converts octal to decimal

mweinelt commented:
> Write a function that converts octal to decimal

This happened. The function name `toIntBase8` sounds weird, but it follows the naming in `lib`.

mweinelt commented:
I have absolutely no clue, why we now have null values in the resulting configuration. In my opinion, the recursive filtering should deal with that, but it does not.

mweinelt commented:
The problem is apparently, that `filterAttrsRecursive` does not recurse into a list.

999eagle commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -195,15 +202,16 @@ in {
     listenerType = workerContext: types.submodule {
```
```suggestion
    listenerType = workerContext: types.submodule ({config, ...}: {
```

mweinelt replied:
I fixed the filtering instead.

999eagle replied:
This is still required to set the correct defaults. Without this we can't set defaults based on `config.path`.
You can use `nix-build -A nixosTests.matrix-synapse.nodes.serverpostgres.system.build.toplevel; grep -o '/nix/store/.*-homeserver\.yaml' result/etc/systemd/system/matrix-synapse.service | xargs cat` to see that the default listener for synapse still contains bind addresses, port *and* mode despite path being null.

mweinelt replied:
Yeah, I noticed. Will be in the next update.

999eagle commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -219,6 +227,34 @@ in {
+
+        mode = mkOption {
+          type = types.nullOr (types.strMatching "^[0,2-7]{3,4}$");
+          default = if path != null then "1660" else null;
```
```suggestion
          default = if config.path != null then "660" else null;
```

This should correctly set `null` as the mode when no `path` is configured.

999eagle commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -244,11 +280,17 @@ in {
         x_forwarded = mkOption {
           type = types.bool;
-          default = false;
+          default = path != null;
```
```suggestion
          default = config.path != null;
```

999eagle replied:
This one is still missing

999eagle commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -105,6 +105,13 @@ let
       };
     });
+
+  toIntBase8 = str:
+    lib.pipe str [
```
```suggestion
  toIntBase8 = str:
    if str == null then null else
    lib.pipe str [
```

Without this, `apply` for the `mode` will fail with "cannot coerce null to a string".

mweinelt replied:
Left the `toIntBase8` function as is, and wrapped `toDecimalFilePermission` around it.

999eagle reviewed (commented):
This should at least fix the listener to have the correct `null` values.

999eagle commented:
We could also replace the filtering of `finalSettings` in line 10 with something like this to correctly remove the `null` values from the generated config file:

```nix
  finalSettings = let
    filter = lib.filterAttrsRecursive (_: v: v != null) cfg.settings;
  in
    filter //
    {
      listeners = builtins.map (l: lib.filterAttrsRecursive (_: v: v != null) l) filter.listeners;
    };
```

mweinelt committed:
nixos/matrix-synapse: fix recursive filtering of null values

Using `filterAttrsRecursive` is not sufficient to account for a nested
attribute set with list values, like used for listeners.

mweinelt commented:
> We could also replace the filtering of finalSettings in line 10

I came up with a recursive variant, that should work.

mweinelt commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -992,6 +1025,17 @@ in {
+        assertion = lib.all
+          (
+            listener:
+              listener.path != null || (listener.bind_addresses != null && listener.port != null)
```
PTAL.

mweinelt commented:
I think I addressed all the feedback. Ready for another review.

999eagle commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -234,17 +293,29 @@ in {
+          else
+            !workerContext;
+          defaultText = ''
+            Enabled, unless the listener is the main instance listener, or it is configured with a UNIX domain socket path.
```
```suggestion
            Enabled for the main instance listener unless it is configured with a UNIX domain socket path.
```

Worker listeners have TLS disabled by default, the main instance listener enables TLS.

mweinelt replied:
Applied

999eagle commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -234,17 +293,29 @@ in {
-          default = false;
+          default = config.path != null;
+          defaultText = ''
+            Enabled, unless the listener is configured with a UNIX domain socket path
```
```suggestion
            Enabled if the listener is configured with a UNIX domain socket path
```

mweinelt replied:
Applied

999eagle commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -1022,15 +1096,26 @@ in {
-    ];
+    ]
+    ++ (map (listener: {
+      assertion = listener.path == null || (listener.path != null && (listener.bind_addresses == null && listener.port == null && listener.tls == null));
```
Isn't the `path != null` check redundant here?

mweinelt replied:
Yes, they are. I reordered the conditions a bit and apparently failed to clean them up.

999eagle commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -1022,15 +1096,26 @@ in {
+      '';
+    }) cfg.settings.listeners)
+    ++ (map (listener: {
+      assertion = listener.path == null || (listener.path != null && listener.type == "http");
```
And this one too?

999eagle reviewed (commented):
Just a few documentation changes. Otherwise LGTM, I'm running this on my server now and it seems to work!

I'm still not entirely happy to see `mode: 432` in my homeserver.yaml without any comment as to what this value means but I guess this isn't something we can really fix in any good way, so I guess we can keep it like that.

mweinelt commented:
@ofborg test matrix-synapse
@ofborg test matrix-synapse-workers

SuperSandro2000 commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -219,6 +250,34 @@ in {
+
+            ::: {.note}
+              This option is incompatible with {option}`bind_addresses`, {option}`port`, {option}`tls`
+              and also does not support the `metrics` and `manhole` listener {option}`type`.
```
```suggestion
              and also does not support the `metrics` and `manhole` listener {option}`type`.
            :::
```
Those shouldn't be optional

SuperSandro2000 commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -234,17 +293,29 @@ in {
           description = lib.mdDoc ''
             Whether to enable TLS on the listener socket.
+
+            ::: {.note}
+              This option will be ignored for UNIX domain sockets.
```
```suggestion
            ::: {.note}
              This option will be ignored for UNIX domain sockets.
            :::
```

Ma27 commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -1022,15 +1096,26 @@ in {
-    ];
+    ]
+    ++ (map (listener: {
+      assertion = listener.path == null || (listener.bind_addresses == null && listener.port == null && listener.tls == null);
```
This accepts a listener with everything being `null` whcih seems wrong:

```
{    path = null;
    bind_addresses = null;
    port = null;
    tls = null;
}
```

mweinelt replied:
I'm stuck at fixing the assertion. See the last fixup commit.

```
trace: null

[...]

       error:
       Failed assertions:
       - Listener configured with UNIX domain socket () ignores the following options:
         - bind_addresses=127.0.0.1 ::1
         - port=8448
         - tls=1
`

The `listener.path` traces as `null`, and the assertion checks for `listener.path != null`.

Ma27 commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -1022,15 +1098,35 @@ in {
+      '';
+    }) cfg.settings.listeners)
+    ++ (map (listener: {
+      assertion = listener.path != null -> (listener.bind_addresses == null || listener.port == null || listener.tls == null);
```
I think this needs to be `listener.path != null -> (listener.bind_addresses == null && ... && ...)`.
Right now, the assertion is still `true` if `path != null` and `port == null`, but `bind_addresses` isn't null.

Ma27 commented:
So I just tried to replace the `http` client/federation listener with a UDS (it's reverse-proxied by nginx anyways) and this fails with `error: value is null while a Boolean was expected` while evaluating `matrix-synapse-register_new_matrix_user`.

Ma27 commented:
Replication via uds rather than ip appears to work though :+1:

mweinelt commented:
> So I just tried to replace the `http` client/federation listener with a UDS (it's reverse-proxied by nginx anyways) and this fails with `error: value is null while a Boolean was expected` while evaluating `matrix-synapse-register_new_matrix_user`.

:scream:

mweinelt commented:
The script has not support for UNIX domain sockets. It would need to be extended with https://github.com/msabramo/requests-unixsocket to support the `http+unix://` proto.

mweinelt commented:
So we now drop the registration script, if the clientListener has no bind addresses.

Ma27 commented:
> Two new options lead me into the shittiest rabbit holes in a long time. It's barely funny anymore.

That's the third escalating PR against Matrix from you that I'm reviewing, so perhaps it's my fault after all. In that case I'm sorry %) 

> So we now drop the registration script, if the clientListener has no bind addresses. 🤷

That's fine for me, I'm using OIDC anyways.
I'm happy to re-check the current state now.

However, I think it's a bit surprising if the script disappears when using UDS, so we should do e.g.
* warn about this in the manual: iirc the registration script is mentioned there anyways.
* perhaps even raise a warning? I.e. `enableRegistrationScript` is `true` by default. If the client listener uses UDS and the option is still true, a warning should be raised.

I'm fine with either option.

mweinelt commented:
> perhaps even raise a warning? I.e. enableRegistrationScript is true by default. If the client listener uses UDS and the option is still true, a warning should be raised.

As long as this leads to an evaluation error, this should probably be a hard assertion. WDYT?

mweinelt commented:
Updated the manual and added both an option and an assertion in the last iteration.
